# Chapter 8

# Functions

### 8.1 Definition

In [3]:
def square(x): 
    return x**2

In [4]:
print(square(2))

4


Hàm square(x) tách phần tính toán bình phương thành một khối riêng, giúp code dễ đọc, dễ tái sử dụng. return sẽ cho phép lấy kết quả của hàm để dùng ở các phần khác trong chương trình. Gọi hàm bằng square(2) sẽ thực hiện tính toán với tham số cụ thể.

In [5]:
print(square(-3))

9


Hàm square(x) tính bình phương của số x và trả kết quả in ra màn hình.

#### 8.1.1 Adding a Description

In [7]:
def square(x):
    """returns the squared value of x""" 
    return x**2

Hàm này tính bình phương của một số và trả kết quả, docstring giúp giải thích chức năng hàm.

#### 8.1.2 Arguments of a Function

In [10]:
def production(l, k, m): 
    """
    Returns the value of the production according to
    labour, capital and materials
    
    Keyword arguments:
    l -- labour (float)
    k -- capital (float)
    m -- materials (float)
    """
    return l**0.3 * k**0.5 * m**(0.2)

Hàm này tính giá trị sản xuất dựa trên lao động, vốn và nguyên liệu và trả kết quả, docstring giúp mô tả rõ ràng mục đích và tham số của hàm.

##### 8.1.2.1 Call Without Argument Names

In [11]:
prod_val = production(60, 42, 40) 
print(prod_val)

46.289449781254994


Đây là cách gọi hàm với các tham số cụ thể, lưu kết quả vào biến và in ra, giúp tái sử dụng kết quả cho các tính toán tiếp theo.

##### 8.1.2.2 Positional Arguments, Arguments by Keywords

In [12]:
def production_2(l, m, k=42): 
    """
    Returns the value of the production according to
    labour, capital and materials
    
    Keyword arguments:
    l -- labour (float)
    m -- materials (float)
    k -- capital (float) (default 42) 
    """
    return l**0.3 * k**0.5 * m**(0.2)

Hàm này tính giá trị sản xuất dựa trên lao động, nguyên liệu và vốn, trong đó tham số vốn có thể bỏ qua để dùng giá trị mặc định, giúp hàm linh hoạt hơn.

In [14]:
# By naming all argument, by ommitting k
prod_val_1 = production_2(l = 42, m = 40)
# By naming all argument and specifying k 
prod_val_2 = production_2(l = 42, m = 40, k = 42) 
# By naming only the argument k
prod_val_3 = production_2(42, 40, k = 42)
# Without naming any argument
prod_val_4 = production_2(42, 40, 42)
res = [prod_val_1, prod_val_2, prod_val_3, prod_val_4] 
print(res)

[41.59215573604822, 41.59215573604822, 41.59215573604822, 41.59215573604822]


Đây là cách gọi hàm linh hoạt trong Python bằng tên, bằng vị trí, kết hợp cả hai, và sử dụng tham số mặc định.

In [1]:
def production_3(a, l, m = 40, k=42): 
    """
    Returns the value of the production according to
    labour, capital and materials
    
    Keyword arguments:
    a -- total factor productivity (float) l -- labour (float)
    m -- materials (float) (default 40)
    k -- capital (float) (default 42)
    """
    return a * l**0.3 * k**0.5 * m**(0.2)

prod_val_1 = production_3(1, 42, m = 38) 
prod_val_2 = production_3(a = 1, l = 42) 
prod_val_3 = production_3(l = 42, a = 1) 
prod_val_4 = production_3(m = 40, l = 42, a = 1)

res = [prod_val_1, prod_val_2, prod_val_3, prod_val_4] 
print(res)

[41.16765711449734, 41.59215573604822, 41.59215573604822, 41.59215573604822]


Đây là cách gọi hàm linh hoạt trong Python bằng tên, bằng vị trí, kết hợp cả hai, và sử dụng tham số mặc định.

##### 8.1.2.3 Function as an Argument to Another Function

In [2]:
def square(x):
    """Returns the squared value of x""" 
    return x**2
    
def apply_fun_to_4(fun):
    """Applies the function `fun` to 4""" 
    return fun(4)
    
print(apply_fun_to_4(square))

16


Minh họa cách truyền một hàm như đối số cho hàm khác. Hàm square tính bình phương một số, còn hàm apply_fun_to_4 nhận một hàm khác và áp dụng nó cho số 4. Khi gọi apply_fun_to_4(square), Python thực hiện square(4) và trả kết quả 16. Ý nghĩa chung là có thể viết các hàm linh hoạt, tái sử dụng cao, áp dụng cho nhiều hàm khác nhau mà không phải viết lại logic.

### 8.2 Scope of a Function

In [3]:
# Definition of a global variable:
value = 1
# Definition of a local variable in function f
def f(x):
    value = 2
    new_value = 3
    print("value equals: ", value) 
    print("new_value equals: ", new_value) 
    return x + value

In [4]:
res = f(3)

value equals:  2
new_value equals:  3


Các biến cục bộ trong hàm (value = 2, new_value = 3) chỉ tồn tại bên trong hàm và dùng để tính toán hoặc in ra trong hàm, trong khi biến toàn cục (value = 1) vẫn giữ nguyên ngoài hàm. Khi gọi hàm, Python dùng giá trị của biến cục bộ để thực hiện phép tính và trả kết quả (res = x + value), nhưng không thay đổi biến toàn cục.

In [5]:
print("value equals: ", value)

value equals:  1


In ra giá trị hiện tại của biến value tại vị trí gọi lệnh. Nếu gọi ngoài hàm, Python sẽ in giá trị biến toàn cục (value = 1). Nếu gọi trong hàm và có biến cục bộ value, Python sẽ in giá trị biến cục bộ (value = 2).

In [6]:
print("new_value equals: ", new_value)

NameError: name 'new_value' is not defined

Các biến cục bộ chỉ dùng được trong hàm, còn ngoài hàm phải dùng biến toàn cục hoặc trả giá trị từ hàm ra nếu muốn sử dụng.

In [7]:
value = 1 
def f(x):
    return x + value 
    
print(f(2))

3


Trong hàm f(x), không có biến cục bộ value, nên Python sử dụng giá trị biến toàn cục. return x + value → 2 + 1 = 3 nên print(f(2)) in ra kết quả 3.

In [9]:
# The value variable is not defined in 
# the local environment of g().
# Python will then search in f(). 
value = 1
def f(): 
    value = 2 
    def g(x):
        return x + value 
    
    return g(2)
print(f())

4


Trong f(), value = 2 → biến cục bộ của hàm f. g(x) không định nghĩa biến value riêng → Python sẽ tìm biến trong môi trường bao quanh (enclosing scope), tức là biến cục bộ value của f(). g(2) → tính 2 + value = 2 + 2 = 4 nên f() trả về g(2) = 4 và print(f()) in ra 4.

In [13]:
# The value variable is not defined in g() or f() 
# but in the higher environment (here, global) 
value = 1
def f():
    def g(x):
        return x + value
    return g(2)
print(f())

3


Khi biến không tồn tại trong hàm hay hàm bao quanh, Python sẽ tìm biến trong phạm vi toàn cục.

In [14]:
def f(x): 
    global y 
    y = x+1

f(3)
print(y)

4


Từ khóa global cho phép tạo hoặc sửa biến toàn cục bên trong hàm. Không dùng global thì biến trong hàm sẽ là cục bộ, không ảnh hưởng biến ngoài hàm.

### 8.3 Lambda Functions

In [15]:
def square(x): 
    return x**2

In [16]:
square_2 = lambda x: x**2 
print(square_2(4))

16


lambda cho phép tạo hàm ngắn gọn trong một dòng mà không cần dùng def. gán hàm này cho biến square_2, để có thể gọi như hàm bình thường. square_2(4) → tính 4**2 = 16.

In [17]:
def production(l, k, m): 
    """
    Returns the value of the production according to
    labour, capital and materials.
    
    Keyword arguments:
    l -- labour (float)
    k -- capital (float)
    m -- materials (float)
    """
    return l**0.3 * k**0.5 * m**(0.2)

In [18]:
production_2 = lambda l,k,m : l**0.3 * k**0.5 * m**(0.2) 
print(production(42, 40, 42))

40.987803063838406


Python cho phép định nghĩa hàm bằng def hoặc lambda. Cả hai cách đều có thể tính toán và trả kết quả giống nhau, nhưng def thuận tiện khi muốn viết docstring hoặc thân hàm phức tạp, còn lambda tiện cho hàm ngắn gọn, dùng một lần.

In [19]:
print(production_2(42, 40, 42))

40.987803063838406


production_2 là hàm lambda tính giá trị sản xuất theo công thức Cobb-Douglas production = 𝑙^0.3 x 𝑘^0.5 x 𝑚^0.2. Gọi production_2(42, 40, 42) → tính giá trị với l=42, k=40, m=42.

### 8.4 Returning Several Values

In [20]:
import statistics 
def desc_stats(x):
    """Returns the mean and standard deviation of `x`"""
    return {"mean": statistics.mean(x), 
            "std_dev": statistics.stdev(x)}
    
x = [1,3,2,6,4,1,8,9,3,2] 
res = desc_stats(x) 
print(res)

{'mean': 3.9, 'std_dev': 2.8460498941515415}


import statistics → nhập module hỗ trợ các phép toán thống kê.

desc_stats(x) → hàm nhận một danh sách x, trả về dictionary gồm:

"mean": trung bình cộng của các giá trị.

"std_dev": độ lệch chuẩn (standard deviation).

Gọi hàm với danh sách x = [1,3,2,6,4,1,8,9,3,2] → tính mean = 3.9, std_dev ≈ 2.846 vàin ra kết quả.

In [21]:
message = "The average value equals {} and the standard deviation is {}" 
print(message.format(res["mean"], res["std_dev"]))

The average value equals 3.9 and the standard deviation is 2.8460498941515415


Sử dụng str.format() để định dạng chuỗi và in kết quả thống kê một cách dễ đọc.

### 8.5 Exercise

1. Create a function named sum_n_integers which returns the sum of the first integer n. Its only argument will be n.

In [22]:
def sum_n_integers(n):
    """
    Trả về tổng của n số nguyên đầu tiên
    """
    return n * (n + 1) // 2

print(sum_n_integers(5))  
print(sum_n_integers(10)) 

15
55


2. Using a loop, display the sum of the first 2 integers, then 3 first integers, then 4 first integers, etc. up to 10.

In [23]:
# Sử dụng vòng lặp để tính tổng dần từ 2 đến 10 số nguyên đầu tiên
for n in range(2, 11):  
    total = sum(range(1, n+1))  
    print(f"Sum of the first {n} integers is {total}")

Sum of the first 2 integers is 3
Sum of the first 3 integers is 6
Sum of the first 4 integers is 10
Sum of the first 5 integers is 15
Sum of the first 6 integers is 21
Sum of the first 7 integers is 28
Sum of the first 8 integers is 36
Sum of the first 9 integers is 45
Sum of the first 10 integers is 55


3. Create a function that from two points represented by pairs of coordinates (x1, y1) and (x2, y2) returns the Euclidean distance between these two points.
Propose a second solution using a lambda function.

In [25]:
import math

def euclidean_distance(p1, p2):
    """
    Trả về khoảng cách Euclid giữa hai điểm p1 và p2
    p1, p2: tuple chứa tọa độ (x, y)
    """
    x1, y1 = p1
    x2, y2 = p2
    return math.sqrt((x2 - x1)**2 + (y2 - y1)**2)

print(euclidean_distance((1, 2), (4, 6)))

5.0


# Kết thúc Chapter 8